src/runtimes/node.js

```diff
@@ -1,38 +1,92 @@
-const { join, basename } = require('path')
+const { basename, dirname, join } = require('path')
 
-const cpFile = require('cp-file')
+const commonPathPrefix = require('common-path-prefix')
 
+const { bundleJsFile } = require('../bundler')
+const { getDependencies, listFilesUsingLegacyBundler } = require('../node_dependencies')
 const { zipNodeJs } = require('../zip_node')
 
-const zipJsFunction = async function ({
-  srcPath,
+const getSrcFiles = function ({ srcPath, mainFile, srcDir, stat, pluginsModulesPath, useEsbuild, externalModules }) {
+  if (!useEsbuild) {
+    return listFilesUsingLegacyBundler({ srcPath, mainFile, srcDir, stat, pluginsModulesPath })
+  }
+
+  if (externalModules.length !== 0) {
+    return getDependencies(mainFile, srcDir, pluginsModulesPath, externalModules)
+  }
+
+  return []
+}
+
+const zipFunction = async function ({
   destFolder,
-  mainFile,
-  filename,
   extension,
-  srcFiles,
+  externalModules,
+  filename,
+  ignoredModules,
+  mainFile,
   pluginsModulesPath,
+  srcDir,
+  srcPath,
+  stat,
   useEsbuild,
-  externalModules,
 }) {
-  if (extension === '.zip') {
-    const destPath = join(destFolder, filename)
-    await cpFile(srcPath, destPath)
+  const destPath = join(destFolder, `${basename(filename, extension)}.zip`)
+  const srcFiles = await getSrcFiles({
+    stat,
+    mainFile,
+    extension,
+    srcPath,
+    srcDir,
+    pluginsModulesPath,
+    useEsbuild,
+    externalModules,
+  })
+  const dirnames = srcFiles.map(dirname)
+
+  if (!useEsbuild) {
+    await zipNodeJs({
+      basePath: commonPathPrefix(dirnames),
+      destFolder,
+      destPath,
+      filename,
+      mainFile,
+      pluginsModulesPath,
+      srcFiles,
+    })
+
     return destPath
   }
 
-  const destPath = join(destFolder, `${basename(filename, extension)}.zip`)
+  const { bundlePath, cleanTempFiles } = await bundleJsFile({
+    additionalModulePaths: pluginsModulesPath ? [pluginsModulesPath] : [],
+    destFilename: filename,
+    destFolder,
+    externalModules,
+    ignoredModules,
+    srcFile: mainFile,
+  })
+
+  // We're adding the bundled file to the zip, but we want it to have the same
+  // name and path as the original, unbundled file. For this, we use a rename.
+  const renames = {
+    [bundlePath]: mainFile,
+  }
+  const basePath = commonPathPrefix([...dirnames, dirname(mainFile)])
+
   await zipNodeJs({
-    srcFiles,
+    basePath,
     destFolder,
     destPath,
     filename,
     mainFile,
     pluginsModulesPath,
-    useEsbuild,
-    externalModules,
+    renames,
+    srcFiles: [...srcFiles, bundlePath],
   })
+  await cleanTempFiles()
```

Comment:
Should this be run in `try`/`finally` in case `zipNodeJs()` throws?

Reply:
Absolutely. That was removed by accident. Reinstated in https://github.com/netlify/zip-it-and-ship-it/pull/325/commits/bff936d6e81ca31e8c4af6dddfd11ff34d0c6fc5.